core/instance.go

```diff
@@ -18,53 +44,255 @@ type instanceTypeInformation struct {
+
+		// checking how many spot instances of this type we already have, so that
+		// we can see how risky it is to launch a new one.
+		spotInstanceCount := i.asg.alreadyRunningSpotInstanceCount(
+			candidate.instanceType, availabilityZone)
+
+		// We skip it in case we have more than 20% instances of this type already
+		// running
+		if spotInstanceCount == 0 ||
+			(*i.asg.DesiredCapacity/spotInstanceCount > 4) {
+			logger.Println(i.asg.name,
+				"no redundancy issues found for", candidate.instanceType,
+				"existing", spotInstanceCount,
+				"spot instances, adding for comparison",
+			)
+
+			filteredInstanceTypes = append(filteredInstanceTypes, candidate.instanceType)
+		} else {
+			logger.Println("\nInstances ", candidate, " and ", existing,
+				"are not compatible or resulting redundancy for the availability zone",
+				"would be dangerously low")
+
+		}
+
+	}
+	logger.Printf("\n Found following compatible instances: %#v\n",
+		filteredInstanceTypes)
+	return filteredInstanceTypes, nil
+
 }
 
-func (it *instance) isSpot() bool {
-	return (it.InstanceLifecycle != nil &&
-		*it.InstanceLifecycle == "spot")
+func (i *instance) getCheapestCompatibleSpotInstanceType() (*string, error) {
+
+	logger.Println("Getting cheapest spot instance compatible to ",
+		*i.InstanceId, " of type", *i.InstanceType)
+
+	filteredInstanceTypes, err := i.getCompatibleSpotInstanceTypes(
+		i.asg.getLaunchConfiguration())
+
+	if err != nil {
+		logger.Println("Couldn't find any compatible instance types", err)
+		return nil, err
+	}
+
+	minPrice := math.MaxFloat64
+	var chosenInstanceType string
+
+	for _, instanceType := range filteredInstanceTypes {
+		price := i.typeInfo.pricing.spot[*i.Placement.AvailabilityZone]
+
+		if price < minPrice {
+			minPrice, chosenInstanceType = price, instanceType
+			logger.Println(i.InstanceId, "changed current minimum to ", minPrice)
+		}
+		logger.Println(i.InstanceId, "cheapest instance type so far is ",
+			chosenInstanceType, "priced at", minPrice)
+	}
+
+	if chosenInstanceType != "" {
+		logger.Println("Chose cheapest instance type", chosenInstanceType)
+		return &chosenInstanceType, nil
+	}
+	logger.Println("Couldn't find any cheaper spot instance type")
+	return nil, fmt.Errorf("No cheaper spot instance types could be found")
+
 }
 
-func (it *instance) terminate(svc *ec2.EC2) {
+func (i *instance) tag(tags []*ec2.Tag) {
 
-	if _, err := svc.TerminateInstances(&ec2.TerminateInstancesInput{
-		InstanceIds: []*string{it.InstanceId},
-	}); err != nil {
-		logger.Println(err.Error())
+	if len(tags) == 0 {
+		logger.Println(i.region.name, "Tagging spot instance", *i.InstanceId,
+			"no tags were defined, skipping...")
+		return
 	}
+
+	svc := i.region.services.ec2
+	params := ec2.CreateTagsInput{
+		Resources: []*string{i.InstanceId},
+		Tags:      tags,
+	}
+
+	logger.Println(i.region.name, "Tagging spot instance", *i.InstanceId)
+
+	for _, err := svc.CreateTags(&params); err != nil; _, err = svc.CreateTags(&params) {
```

Comment:
Even though I wrote this not so long time ago, I also find it hard to understand. This code is a bit write-only and may need some refactoring :)

What is does is trying to tag the instance in a loop until the tagging returns no error, since the instance has just been launched and some API calls executed against it may be failing for a short while immediately after launch.